src/github_permission_manager_demotion/handler.py

```diff
@@ -0,0 +1,127 @@
+        self.installation_id = self.get_ssm_parameter(f"/github_permission_manager_webhook/${workspace}_installation_id")
+        self.auth_headers = self.get_token_to_access_github()
+
+    def get_token_to_access_github(self):
+        github_auth = GitHubAuth(self.private_key, self.app_id, self.installation_id)
+        access_token = github_auth.get_access_token()
+        return {
+            "Authorization": f"Bearer {access_token}",
+            "Accept": "application/vnd.github.v3+json"
+        }
+
+    def get_ssm_parameter(self, parameter_name):
+        parameter = self.ssm_client.get_parameter(Name=parameter_name, WithDecryption=True)
+        return parameter['Parameter']['Value']
+
+    def close_issue(self, repository, issue_number):
+        response = self.requests.patch(
+            f"https://api.github.com/repos/{repository}/issues/{issue_number}",
+            headers=self.auth_headers,
+            json={"state": "closed"}
+        )
+        print(f"Issue closed: {response.status_code}")
+
+    def post_comment_on_issue(self, repository, issue_number, comment_body):
+        data = {
+            "body": comment_body
+        }
+        response = requests.post(
+            f"https://api.github.com/repos/{repository}/issues/{issue_number}/comments",
+            headers=self.auth_headers,
+            json=data
+        )
+        print(f"Comment posted: {response.status_code}")
+
+    def get_all_org_owners(self, organization):
+        response = requests.get(
+            f"https://api.github.com/orgs/{organization}/members?role=admin",
+            headers=self.auth_headers
+        )
+        print(f"Request to get org owners: {response.status_code}")
+        org_owners = []
+        for owner in response.json():
+            print(f"Owner: {owner['login']}")
+            org_owners.append(owner['login'])
+        return org_owners
+
+
+    def make_member_on_github(self, organization, user):
+        data = {
+            "role": "member"
+        }
+        response = requests.put(
+            f"https://api.github.com/orgs/{organization}/memberships/{user}",
+            headers=self.auth_headers,
+            json=data
+        )
+        print(f"Comment posted: {response.status_code}")
+
+    def demote_user_lambda(self, event, _context):
+        self.initialise_aws_clients()
+        self.get_all_parameters()
+        organization = event.get('organization')
+        user = event.get('user')
+        repository = event.get('repository')
+        issue_number = event.get('issue_number')
+        org_owners = self.get_all_org_owners(organization)
+        if user in org_owners:
+            if is_last_org_owner(org_owners, user):
+                self.post_comment_on_issue(repository, issue_number, "User is the last owner - therefore will not be demoted")
+                self.close_issue(repository, issue_number)
+                return
```

Comment:
I feel like maybe there should be some more alarms bells on this, maybe something that posts to teams?

Reply:
I'd suggest this is something for a future story to define what should happen here. For the initial work and running in a test organisation I'd say this is OK, but before taking any further we should work through the edge cases.